src/coreclr/src/gc/gc.cpp

```diff
@@ -20636,9 +20636,26 @@ void gc_heap::mark_phase (int condemned_gen_number, BOOL mark_only_p)
     if (gc_t_join.joined())
 #endif //MULTIPLE_HEAPS
     {
+        uint64_t promoted_bytes = 0;
 #ifdef HEAP_ANALYZE
         heap_analyze_enabled = FALSE;
-        GCToEEInterface::AnalyzeSurvivorsFinished(condemned_gen_number);
+#ifdef MULTIPLE_HEAPS
+        for (int i = 0; i < n_heaps; i++)
+        {
+            promoted_bytes += g_promoted[i * 16];
+        }
+#else
+        promoted_bytes = g_promoted;
+#endif //MULTIPLE_HEAPS
+
+        GCToEEInterface::AnalyzeSurvivorsFinished(condemned_gen_number, promoted_bytes, [](){
+            g_theGCHeap->DiagDescrGenerations([](void*, int generation, uint8_t* rangeStart, uint8_t* rangeEnd, uint8_t* rangeEndReserved)
+            {
+                uint64_t range = static_cast<uint64_t>(rangeEnd - rangeStart);
+                uint64_t rangeReserved = static_cast<uint64_t>(rangeEndReserved - rangeStart);
+                FIRE_EVENT(GCGenerationRange, generation, rangeStart, range, rangeReserved);
```

Comment:
can we take this opportunity to actually make this a named function and use it here and in gcee.cpp so we don't duplicate this code?